proxy/api/src/patch.rs

```diff
@@ -0,0 +1,144 @@
+    pub commit: Oid,
+    /// The merge base of [`Patch::commit`] and the head commit of the first maintainer's default
+    /// branch.
+    pub merge_base: Option<Oid>,
+}
+
+/// List all patches for the given project.
+///
+/// # Errors
+/// * Cannot access the monorepo
+/// * Cannot find references within the monorepo
+pub async fn list(
+    peer: &net::peer::Peer<BoxedSigner>,
+    project_urn: Urn,
+) -> Result<Vec<Patch>, crate::error::Error> {
+    let mut patches = Vec::new();
+
+    let default_branch_head_commit_id = {
+        let project = state::get_project(peer, project_urn.clone())
+            .await?
+            .ok_or_else(|| state::Error::ProjectNotFound(project_urn.clone()))?;
+        let maintainer = project
+            .delegations()
+            .iter()
+            .flat_map(|either| match either {
+                Either::Left(pk) => Either::Left(std::iter::once(pk)),
+                Either::Right(indirect) => Either::Right(indirect.delegations().iter()),
+            })
+            .next()
+            .expect("missing delegation");
+        let default_branch = state::get_branch(
+            peer,
+            project_urn.clone(),
+            Some(PeerId::from(*maintainer)),
+            None,
+        )
+        .await?;
+        crate::browser::using(peer, default_branch, move |browser| {
+            Ok(browser.get().first().clone())
+        })
+        .await?
+        .id
+    };
+
+    for project_peer in state::list_project_peers(peer, project_urn.clone()).await? {
+        let remote = match &project_peer {
+            project::Peer::Local { .. } => None,
+            project::Peer::Remote { peer_id, .. } => Some(*peer_id),
+        };
+
+        let ref_scope = match remote {
+            Some(remote) => RefScope::Remote {
+                name: Some(remote.to_string()),
+            },
+            None => RefScope::Local,
+        };
+
+        let branch = match state::get_branch(peer, project_urn.clone(), remote, None).await {
+            Ok(branch) => branch,
+            Err(state::Error::MissingRef { .. }) => {
+                // The peer hasn’t published any branches yet.
+                continue;
+            },
+            Err(e) => return Err(e.into()),
+        };
+
+        let identity = project_peer
+            .clone()
+            .replicated()
+            .map(|peer| identity::Identity::from((peer.peer_id(), peer.status().user.clone())));
+        let peer_id = project_peer.peer_id();
+
+        crate::browser::using(peer, branch, {
+            let patches = &mut patches;
+            move |browser| {
+                let tags = browser.list_tags(ref_scope)?;
+                for tag in tags {
+                    match tag {
+                        radicle_source::surf::git::Tag::Light { .. } => {
+                            continue;
+                        },
+                        radicle_source::surf::git::Tag::Annotated {
```

Comment:
Using the suggestion of splitting `Patch` data, you could define a smart constructor `fn patch(tag: surf::Tag) -> Option<Patch>` and use a `filter_map` here.

Reply:
I’ve tried it but because of errors and access to `browser` this is call-site code I came up with to replace the for-loop.
```rust
Ok(patches.extend(
    browser
        .list_tags(ref_scope)?
        .into_iter()
        .filter_map(|tag| tag_to_patch(browser, tag).transpose())
        .collect::<Result<Vec<Patch>, _>>()?,
))
```
And `tag_to_patch` would be pretty much the body of the for-loop. This looks much harder to understand and is probably less performant because it allocates. So iteration seems fine to me.

Reply:
Whatever you found best works :)